demos/guide/quic-server-ssl-poll-http.c

```diff
@@ -0,0 +1,1953 @@
+        ossl_list_pe_insert_head(&pm->pm_head, pe);
+        pm->pm_need_rebuild = 1;
+        pe->pe_my_pm = pm;
+    }
+}
+
+static void
+remove_pe_from_pm(struct poll_manager *pm, struct poll_event *pe)
+{
+    if (pe->pe_my_pm == pm) {
+        ossl_list_pe_remove(&pm->pm_head, pe);
+        pm->pm_need_rebuild = 1;
+        pe->pe_my_pm = NULL;
+    }
+}
+
+static struct poll_manager *
+create_poll_manager(void)
+{
+    struct poll_manager *pm = NULL;
+
+    pm = OPENSSL_zalloc(sizeof (struct poll_manager));
+    if (pm == NULL)
+        return NULL;
+
+    ossl_list_pe_init(&pm->pm_head);
+    pm->pm_poll_set = OPENSSL_malloc(sizeof (struct poll_event) * POLL_GROW);
+    if (pm->pm_poll_set != NULL) {
+        pm->pm_poll_set_sz = POLL_GROW;
+        pm->pm_event_count = 0;
+    } else {
+        OPENSSL_free(pm);
+    }
+
+    return pm;
+}
+
+static int
+rebuild_poll_set(struct poll_manager *pm)
+{
+    struct poll_event *new_poll_set;
+    struct poll_event *pe;
+    size_t new_sz;
+    size_t pe_num;
+    size_t i;
+
+    if (pm->pm_need_rebuild == 0)
+        return 0;
+
+    pe_num = ossl_list_pe_num(&pm->pm_head);
+    if (pe_num > pm->pm_poll_set_sz) {
+        /*
+         * grow poll set by POLL_GROW
+         */
+        new_sz = sizeof (struct poll_event) * (pm->pm_poll_set_sz + POLL_GROW);
+        new_poll_set = (struct poll_event *)OPENSSL_realloc(pm->pm_poll_set,
+                                                            new_sz);
+        if (new_poll_set == NULL)
+            return -1;
+        pm->pm_poll_set = new_poll_set;
+        pm->pm_poll_set_sz += POLL_GROW;
+
+    } else if ((pe_num + POLL_DOWNSIZ) < pm->pm_poll_set_sz) {
+        /*
+         * shrink poll set by POLL_DOWNSIZ
+         */
+        new_sz = sizeof (struct poll_event) *
+            (pm->pm_poll_set_sz - POLL_DOWNSIZ);
+        new_poll_set = (struct poll_event *)OPENSSL_realloc(pm->pm_poll_set,
+                                                            new_sz);
+        if (new_poll_set == NULL)
+            return -1;
+        pm->pm_poll_set = new_poll_set;
+        pm->pm_poll_set_sz -= POLL_GROW;
+    }
+
+    i = 0;
+    DPRINTF(stderr, "%s there %zu events to poll\n", __func__,
+            ossl_list_pe_num(&pm->pm_head));
+    OSSL_LIST_FOREACH(pe, pe, &pm->pm_head) {
+        pe->pe_poll_item.events = pe->pe_want_events;
+        pm->pm_poll_set[i++] = *pe;
+        DPRINTF(stderr, "\t%p (%s) " POLL_FMT " (disabled: " POLL_FMT ")\n",
+                pe, pe_type_to_name(pe),
+                POLL_PRINTA(pe->pe_poll_item.events),
+                POLL_PRINTA(~pe->pe_want_mask));
+    }
+    pm->pm_event_count = i;
+    pm->pm_need_rebuild = 0;
+
+    return 0;
+}
+
+static void
+destroy_poll_manager(struct poll_manager *pm)
+{
+    struct poll_event *pe, *pe_safe;
+
+    if (pm == NULL)
+        return;
+
+    OSSL_LIST_FOREACH_DELSAFE(pe, pe_safe, pe, &pm->pm_head) {
+        SSL_free(get_ssl_from_pe(pe));
+        destroy_pe(pe);
```

Comment:
Should these two calls be reversed?  i.e. when the callback in destroy_pe is called, do we want the SSL associated with the poll_entry to still be valid so the callback can operate on it?

Reply:
`SSL_free()` was wrong, we are supposed to free the wrapped `ssl` object inside `destroy_pe()`